core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -70,18 +72,225 @@ class SYCLInternal {
 
   std::unique_ptr<sycl::queue> m_queue;
 
+  // USMObjectMem is a reusable buffer for a single object
+  // in USM memory
+  template <sycl::usm::alloc Kind>
+  class USMObjectMem {
+   public:
+    class Deleter {
+     public:
+      Deleter() = default;
+      explicit Deleter(USMObjectMem* mem) : m_mem(mem) {}
+
+      template <typename T>
+      void operator()(T* p) const noexcept {
+        assert(m_mem);
+        assert(sizeof(T) == m_mem->m_size);
+
+        if constexpr (sycl::usm::alloc::device == kind)
+          // Only skipping the dtor on trivially copyable types
+          static_assert(std::is_trivially_copyable_v<T>);
+        else
+          p->~T();
+
+        m_mem->m_size = 0;
+      }
+
+     private:
+      USMObjectMem* m_mem = nullptr;
+    };
+
+    static constexpr sycl::usm::alloc kind = Kind;
+
+    void reset() {
+      assert(m_size == 0);
+
+      if (m_data) {
+        sycl::free(m_data, *m_q);
+        m_capacity = 0;
+        m_data     = nullptr;
+      }
+      m_q.reset();
+    }
+
+    void reset(sycl::queue q) {
+      reset();
+      m_q.emplace(std::move(q));
+    }
+
+    USMObjectMem() = default;
+    explicit USMObjectMem(sycl::queue q) noexcept : m_q(std::move(q)) {}
+
+    USMObjectMem(USMObjectMem const&) = delete;
+    USMObjectMem(USMObjectMem&&)      = delete;
+    USMObjectMem& operator=(USMObjectMem&&) = delete;
+    USMObjectMem& operator=(USMObjectMem const&) = delete;
+
+    ~USMObjectMem() {
+      assert(m_size == 0);
+
+      if (m_data) sycl::free(m_data, *m_q);
+    }
+
+    void* data() noexcept { return m_data; }
+    const void* data() const noexcept { return m_data; }
+
+    size_t size() const noexcept { return m_size; }
+    size_t capacity() const noexcept { return m_capacity; }
+
+    // reserve() allocates space for at least n bytes
+    // returns the new capacity
+    size_t reserve(size_t n) {
+      assert(m_size == 0);
+      assert(m_q);
+
+      if (m_capacity < n) {
+        // First free what we have (in case malloc can reuse it)
+        sycl::free(m_data, *m_q);
+
+        m_data = sycl::malloc(n, *m_q, kind);
+        if (!m_data) {
+          m_capacity = 0;
+          throw std::bad_alloc();
+        }
+
+        m_capacity = n;
+      }
+
+      return m_capacity;
+    }
+
+   private:
+    static void fence(sycl::event& wat) {
+      try {
+        wat.wait_and_throw();
+      } catch (sycl::exception const& e) {
+        Kokkos::Impl::throw_runtime_exception(
+            std::string("There was a synchronous SYCL error:\n") += e.what());
+      }
+    }
+
+    static void fence(sycl::queue& wat) {
+      try {
+        wat.wait_and_throw();
+      } catch (sycl::exception const& e) {
+        Kokkos::Impl::throw_runtime_exception(
+            std::string("There was a synchronous SYCL error:\n") += e.what());
+      }
+    }
```

Comment:
Hmm.. I'm not quite sure that this is the right place for it. I would rather go for the execution space `Kokkos::Experimental::SYCL`.

Reply:
I'm trying to do that (I agree that is where it belongs), but ran into incomplete types because of include order.

If its okay, I'll just add a FIXME_SYCL here and work on it separately.

Reply:
Yes, this pull request has been dragging long enough. I am in favor of merging and cleaning up when we have merged more of the `SYCL` pull requests.

You still need to rebase apparently, though.